// angular2-cache/memory/MapCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MapCache = (function () {
    function MapCache() {
        this._cache = new Map();
    }
    /**
     * @override
     */
    MapCache.prototype.setCachedValue = function (key, value) {
        this._cache.set(key, value);
    };
    /**
     * @override
     */
    MapCache.prototype.getCachedValue = function (key) {
        return this._cache.get(key);
    };
    /**
     * @override
     */
    MapCache.prototype.clear = function () {
        this._cache.clear();
    };
    /**
     * @override
     */
    MapCache.prototype.size = function () {
        return this._cache.size;
    };
    /**
     * @override
     */
    MapCache.prototype.setEnableLogging = function (enabled) {
        throw Error("UnsupportedException");
    };
    /**
     * @override
     */
    MapCache.prototype.isLoggingEnabled = function () {
        throw Error("UnsupportedException");
    };
    /**
     * @override
     */
    MapCache.prototype.setEnable = function (enabled) {
        throw Error("UnsupportedException");
    };
    return MapCache;
}());
exports.MapCache = MapCache;
