// angular2-cache/pipe/CachedDatePipe.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var cache_1 = require("../decorator/cache");
var CachedDatePipe = (function () {
    function CachedDatePipe(locale) {
        this.datePipe = new common_1.DatePipe(locale.toString());
    }
    return CachedDatePipe;
}());
exports.CachedDatePipe = CachedDatePipe;
var ZoneCachedDatePipe = (function (_super) {
    __extends(ZoneCachedDatePipe, _super);
    function ZoneCachedDatePipe(locale) {
        return _super.call(this, locale) || this;
    }
    /**
     * @override
     */
    ZoneCachedDatePipe.prototype.transform = function (date, pattern) {
        return this.datePipe.transform(date, pattern);
    };
    return ZoneCachedDatePipe;
}(CachedDatePipe));
ZoneCachedDatePipe.decorators = [
    { type: core_1.Pipe, args: [{
                name: 'zoneCachedDate'
            },] },
];
/** @nocollapse */
ZoneCachedDatePipe.ctorParameters = function () { return [
    { type: core_1.OpaqueToken, decorators: [{ type: core_1.Inject, args: [core_1.LOCALE_ID,] },] },
]; };
__decorate([
    cache_1.ZoneCached(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String]),
    __metadata("design:returntype", String)
], ZoneCachedDatePipe.prototype, "transform", null);
exports.ZoneCachedDatePipe = ZoneCachedDatePipe;
var MemoryCachedDatePipe = (function (_super) {
    __extends(MemoryCachedDatePipe, _super);
    function MemoryCachedDatePipe(locale) {
        return _super.call(this, locale) || this;
    }
    /**
     * @override
     */
    MemoryCachedDatePipe.prototype.transform = function (date, pattern) {
        return this.datePipe.transform(date, pattern);
    };
    return MemoryCachedDatePipe;
}(CachedDatePipe));
MemoryCachedDatePipe.decorators = [
    { type: core_1.Pipe, args: [{
                name: 'memoryCachedDate'
            },] },
];
/** @nocollapse */
MemoryCachedDatePipe.ctorParameters = function () { return [
    { type: core_1.OpaqueToken, decorators: [{ type: core_1.Inject, args: [core_1.LOCALE_ID,] },] },
]; };
__decorate([
    cache_1.MemoryCached(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String]),
    __metadata("design:returntype", String)
], MemoryCachedDatePipe.prototype, "transform", null);
exports.MemoryCachedDatePipe = MemoryCachedDatePipe;
