// angular2-cache/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var cache_1 = require("./decorator/cache");
exports.ZoneCached = cache_1.ZoneCached;
exports.MemoryCached = cache_1.MemoryCached;
exports.FileCached = cache_1.FileCached;
exports.SessionCached = cache_1.SessionCached;
exports.StorageCached = cache_1.StorageCached;
var CachedDatePipe_1 = require("./pipe/CachedDatePipe");
exports.ZoneCachedDatePipe = CachedDatePipe_1.ZoneCachedDatePipe;
exports.MemoryCachedDatePipe = CachedDatePipe_1.MemoryCachedDatePipe;
var NgZoneGlobalCache_1 = require("./zone/NgZoneGlobalCache");
exports.NgZoneGlobalCache = NgZoneGlobalCache_1.NgZoneGlobalCache;
var MemoryGlobalCache_1 = require("./memory/MemoryGlobalCache");
exports.MemoryGlobalCache = MemoryGlobalCache_1.MemoryGlobalCache;
var MemoryCache_1 = require("./memory/MemoryCache");
exports.MemoryCache = MemoryCache_1.MemoryCache;
var Utils_1 = require("./Utils");
exports.CacheKeyBuilder = Utils_1.CacheKeyBuilder;
var CacheProviders_1 = require("./CacheProviders");
exports.CACHE_PROVIDERS = CacheProviders_1.CACHE_PROVIDERS;
var CacheModule_1 = require("./CacheModule");
exports.CacheModule = CacheModule_1.CacheModule;
